### PR TITLE
[API] Introduces new "Score Media" API

### PR DESCRIPTION
**Pull request by nkpng2k**
Introduces api for image scoring in openapi 3.0 spec similar to https://github.com/h2oai/mlops-byom-images/pull/63

* moves original api spec to `v1` directory, but preserves build policies and paths (uses openapi v2.0 spec)
* introduces new `v2` api directory which implements openapi v3.0 spec
* introduces additional api endpoint to match above pr 
* bumps several dependencies including gradle (hence @achraf-mer why I'm tagging you)
* implementes empty `UNIMPLEMENTED` endpoint in `local-rest-scorer` 
* adds some tests.

### CODE REVIEW

**Review by achraf-mer — APPROVED** [review]
LGTM, once built we can run a scan to check.

File: common/swagger/v2/swagger.yaml
Comment (orendain) [inline common/swagger/v2/swagger.yaml:62]: So as to not be tied to the evolution of the non-media v1 spec, could this and the response have its own explicit definition?

File: common/swagger/v2/swagger.yaml
Comment (orendain) [inline common/swagger/v2/swagger.yaml:39]: Does this mean that there must be a field labeled `payload` in the POST call?  What is the type of that field?  Is it an object itself, or a JSON string?

Would it make sense to flatten the structure so that those fields are on the same level as `files`?

**Review by orendain — CHANGES_REQUESTED** [review]
Requested changes was for explicit definition of objects in the API (untied from v1 API), but had some other questions.

Thank you @nkpng2k !

File: local-rest-scorer/src/main/java/ai/h2o/mojos/deploy/local/rest/controller/ModelsApiControllerV2.java
Comment (nkpng2k) [reply local-rest-scorer/src/main/java/ai/h2o/mojos/deploy/local/rest/controller/ModelsApiControllerV2.java:27]: oh oops... forgot to delete that comment :sweat_smile:

File: local-rest-scorer/src/main/java/ai/h2o/mojos/deploy/local/rest/converters/ScoreMediaRequestConverter.java
Comment (nkpng2k) [reply local-rest-scorer/src/main/java/ai/h2o/mojos/deploy/local/rest/converters/ScoreMediaRequestConverter.java:12]: slightly different. The `transformers` in the above are related to data transformations as opposed to http request transformations.... I suppose we could have some common http converters, but this is a springboot specific thing so unrelated to the code in `transformers`

File: common/swagger/v2/swagger.yaml
Comment (nkpng2k) [reply common/swagger/v2/swagger.yaml:62]: tbh I debated this as well... in fact, I even considered having a single request spec that both apis inherit from, because in reality it makes much more sense to have the requests and responses be unified and evolve independently of either api spec. Like each api should be dependent on the requests and responses themselves and while the apis can evolve they would be constrained by the requests/responses such that there is some level of consistency.

File: common/swagger/v2/swagger.yaml
Comment (nkpng2k) [reply common/swagger/v2/swagger.yaml:22]: I think `url` can just be `/` I'll try it out.

File: common/swagger/v2/swagger.yaml
Comment (nkpng2k) [reply common/swagger/v2/swagger.yaml:5]: :+1:

File: common/swagger/v2/swagger.yaml
Comment (nkpng2k) [reply common/swagger/v2/swagger.yaml:39]: yes the request would take the format:
```python
res = requests.post("http://localhost:8080/model/score/media", data={"payload": json.dumps(req)}, files=files)
```
using the requests package in python.

Sadly, no this is not possible. Form data does not support complex objects like `list of list` therefore it is not possible to define the payload as independent form fields.

Comment (nkpng2k) [reply common/swagger/v2/swagger.yaml:39]: @orendain

**Comment by nkpng2k** [comment]
@orendain I added new model specifications for v2 api, so now each of the apis are now independent of each other... although at the moment that means a ton of duplicated "code"

File: gradle/mixins/dependencies.gradle
Comment (orendain) [inline gradle/mixins/dependencies.gradle:54]: Didn't notice this the first time.  QQ: why are these needed?

**Review by orendain — COMMENTED** [review]
Performance considerations aside (e.g., whether an underlying scorer can calculate contribution with less compute during scoring than if done separately), WDYT about removing the contributions-related bits from this endpoint?

- There had been some multiple discussions about whether it makes sense to keep scoring and shap endpoints separated.  I can't recall/find arguments for keeping them combined at an API level.
- This v2 API is not necessarily meant to be backward compatible with the v1 Scoring API. Adopting the API from v1 Scoring API makes this API an extension and not be able to serve as the next evolution.  For example, if when deprecating v1, the decision is to separate the concerns above, the endpoint here becomes a conflict.
- Granted, this may raise a feature request to support shapley values for image scoring, which would also come baked if the API looked identical to v1.
- If not, then perhaps we should mark this API not as v2 but as a v1 extension?  And reserve v2 for when APIs are evolved.

WDYT?

---

Is it possible to separate the implementation from the spec?  ~20-25 files are related to the new spec (and moving things around), and ~5 to the implementation in the local rest scorer.  That way, we also hold off from tying the scorer to changes in v2 API, especially if there aren't any plans to do anything other than throw unimplemented.

File: common/swagger/v2/swagger.yaml
Comment (nkpng2k) [reply common/swagger/v2/swagger.yaml:32]: ummmm... @orendain just to be clear `shapley contributions` != `image/media scoring`

sure we could flatten the payload, I guess the question becomes, do we want to validate a single string, or multiple different strings, in the backend, you'd end up having to parse 2 strings using some version of `json.loads()` to get fields which is an array of strings, and rows which is an array of arrays of strings. Personally, I don't like this... like from client call perspective, it increases the amount of work required to construct the proper payload request, and does not (imho) significantly improve the ability to understand or utilize the api.

**Comment by nkpng2k** [comment]
offline discussion with @orendain:
1. extract implementation from api spec pr
2. use more rest compliant endpoint. ex `/model/mediaScore` (? maybe not camel case... check restful spec)
3. not v2, v1exp/v1ext

**Comment by nkpng2k** [comment]
@orendain completed changes as per offline discussion, also reverted all changes associated with implementation in `local-rest-scorer`

**Review by orendain — APPROVED** [review]
Clean new experimental API without affecting any existing binaries!

LGTM. Thank you @nkpng2k!

**Pull request merged**
